### ISSUE TITLE
Functional interfaces targeted in lambdas are imported when using auto-import printer

### ISSUE THREAD

**Issue by slarse** [INRIA/spoon#4079]
To me, that sounds like imports are added for functional interfaces even when they are used as lambda expression (and likely method references too?). Does that match your processed/printed code?

_Originally posted by @SirYwell in https://github.com/INRIA/spoon/issues/4070#issuecomment-893328459_

I've verified this behavior. When a functional interface is used as targeted in a lambda, it's imported by the auto-importer, even if it's never explicitly referenced. Here's a minimal example to reproduce the problem:

```java
// TargetsFunctionalInterface.java
public class TargetsFunctionalInterface {
    public static void main(String[] args) {
        HasFunctionalInterface.supplyOne(i -> System.out.println(i));
    }
}

// HasFunctionalInterface.java
import java.util.function.IntFunction;

public class HasFunctionalInterface {
    public static void supplyOne(IntFunction intFunction) {
        intFunction.apply(1);
    }
}
```

When printing these files again, the line `import java.util.function.IntFunction;` is added to the top of `TargetsFunctionalInterface.java`. The **expected result** is of course that this does not happen.

Note that both `TargetsFunctionalInterface.java` _and_ `HasFunctionalInterface.java` must be added as resources. If only `HasFunctionalInterface.java` is not added, the `IntFunction` reference can't be resolved by Spoon, and consequently the import isn't added.

**Comment by slarse** [comment]
The problem here is that the `ForceImportProcessor` sets the reference of the lambda's type (here `java.util.function.IntFunction`) to be simply qualified, which prompts the `ImportCleaner` to later import that type.

I can see three reasonable solutions.

1. When setting the type of a lambda, always make that type implicit (i.e. `ref.setImplicit(true)` in `CtLambdaImpl.setType(CtTypeReference<?> ref)`
    - Pros: Very easy, and intuitively makes sense as the type of a lambda cannot be explicitly written out in Java.
    - Cons: Possibly unexpected side effect of the `setType` method.
2. Avoid marking the types of `CtLambda` as simply qualified in `ForceImportProcessor`
    - Pros: Relatively easy.
    - Cons: Possibly unintuitive as a corner case of the `ForceImportProcessor`
3. Avoid adding imports for the type of a `CtLambda` in `ImportCleaner`
    - Pros: Relatively easy.
    - Cons: Annoying to have yet another corner case in `ImportCleaner`.

I'm leaning towards option 1.

**Comment by slarse** [comment]
I've implemented option 1 in #4093. I'm not sure that's the way to go, though. @monperrus any thoughts?
